### PR TITLE
ADComputeSmearedCrackingStress

### PR DESCRIPTION
**Pull request by recuero**
This PR converts `ComputeSmearedCrackingStress` to automatic differentiation, together with other used classes, such as `SmearedCrackSofteningBase`.

The approach used here is to create a separate AD version.

Refs #15311 

@lynnmunday

### CODE REVIEW

**Comment by recuero** [comment]
@dschwen please review

File: modules/tensor_mechanics/include/materials/ADAbruptSoftening.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+class ADAbruptSoftening : public ADSmearedCrackSofteningBase
+{
+public:
+  static InputParameters validParams();
+
+  ADAbruptSoftening(const InputParameters & parameters);
+
+  virtual void computeCrackingRelease(ADReal & stress,
+                                      ADReal & stiffness_ratio,
+                                      const ADReal strain,
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ADAbruptSoftening.h:28]: please make these `const ADReal &`. Copying a dual number comes at a significant cost.

File: modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+   * @param cracking_alpha Initial slope of exponential softening model
+   * @param youngs_modulus Young's modulus
+   */
+  virtual void computeCrackingRelease(int i,
+                                      ADReal & sigma,
+                                      ADReal & stiffness_ratio,
+                                      const ADReal strain_in_crack_dir,
+                                      const ADReal cracking_stress,
+                                      const ADReal cracking_alpha,
+                                      const ADReal youngs_modulus);
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h:63]: also here (and elsewhere if applicable)

File: modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  ///@}
+
+  //@{ Damage (goes from 0 to 1) in crack directions
+  ADMaterialProperty<RealVectorValue> & _crack_damage;
+  const MaterialProperty<RealVectorValue> & _crack_damage_old;
+  ///@}
+
+  /// Vector of values going from 1 to 0 as crack damage accumulates. Legacy
+  /// property for backward compatibility -- remove in the future.
+  ADMaterialProperty<RealVectorValue> & _crack_flags;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h:135]: Doesn't look like this is used at all (the AD property is set up, but I do not see anywhere the AD property is actually gotten)

File: modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  ///@}
+
+  //@{ Maximum strain in direction of crack
+  ADMaterialProperty<RealVectorValue> & _crack_max_strain;
+  const MaterialProperty<RealVectorValue> & _crack_max_strain_old;
+  ///@}
+
+  //@{ Variables used by multiple methods within the calculation for a single material point
+  ADRankFourTensor _local_elasticity_tensor;
+  ///@}
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h:154]: `///@` bracketing is not necessary here

File: modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h (hunk 3 above)
Comment (dschwen) [reply modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h:135]: I see, it is output to the exodus file. We might want to consider just removing them and skipping them in the exodiff compare file

File: modules/tensor_mechanics/src/materials/ADComputeSmearedCrackingStress.C [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+      const ADReal c02 = (c0_coupled && c2_coupled ? 1.0 : 0.0);
+      const ADReal c12 = (c1_coupled && c2_coupled ? 1.0 : 0.0);
+
+      const ADReal c01_shear_retention = (c0_coupled && c1_coupled ? 1.0 : _shear_retention_factor);
+      const ADReal c02_shear_retention = (c0_coupled && c2_coupled ? 1.0 : _shear_retention_factor);
+      const ADReal c12_shear_retention = (c1_coupled && c2_coupled ? 1.0 : _shear_retention_factor);
+
+      // Filling with 9 components is sufficient because these are the only nonzero entries
+      // for isotropic or orthotropic materials.
+      std::vector<ADReal> local_elastic(9);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeSmearedCrackingStress.C:346]: creating and destroying std::vectors is costly. I would suggest making this a member variable!

**Review by dschwen — CHANGES_REQUESTED** [review]
Looks great, except for a few minor changes.

File: modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h (hunk 3 above)
Comment (recuero) [reply modules/tensor_mechanics/include/materials/ADComputeSmearedCrackingStress.h:135]: There are tests checking for those flags specifically...

**Review by dschwen — APPROVED** [review]

**Pull request merged**